File: packages/validator-go/validator/expr/ast.go
package expr

// AST node set (EXPRESSION-GRAMMAR §4). Field names follow the JS reference
// (ConditionParser.ts); ToMap() emits the JS serialization the 4-language
// equivalence compares (position-free, GRAMMAR §4 position note).
//
//	Ternary { type:'Ternary', condition, trueValue, falseValue }
//	Binary  { type:'Binary', operator, left, right }
//	Unary   { type:'Unary', operator, operand }
//	In      { type:'In', negated, value, list:[] }
//	Path    { type:'Path', relative, levelsUp, segments:[] }
//	Literal { type:'Literal', valueType, value }
//	Group   { type:'Group', expression }
//
// PathSegment = {type:'identifier', value} | {type:'wildcard'} | {type:'index', value:number}

// Node is one AST node. ToMap renders the JS serialization shape.
type Node interface {
	// ToMap serializes the node (and its children) to the shared-fixture AST shape.
	ToMap() map[string]any
}

// TernaryNode is cond ? trueValue : falseValue (right-associative).
type TernaryNode struct {
	Condition  Node
	TrueValue  Node
	FalseValue Node
}

// ToMap serializes the Ternary node and its children to the shared-fixture
// AST shape {type, condition, trueValue, falseValue}.
func (n *TernaryNode) ToMap() map[string]any {
	return map[string]any{
		"type":       "Ternary",
		"condition":  n.Condition.ToMap(),
		"trueValue":  n.TrueValue.ToMap(),
		"falseValue": n.FalseValue.ToMap(),
	}
}

// BinaryNode is a comparison (== != > >= < <=) or logical (&& ||) operation.
type BinaryNode struct {
	Operator string
	Left     Node
	Right    Node
}

// ToMap serializes the Binary node and its operands to the shared-fixture AST
// shape {type, operator, left, right}.
func (n *BinaryNode) ToMap() map[string]any {
	return map[string]any{
		"type":     "Binary",
		"operator": n.Operator,
		"left":     n.Left.ToMap(),
		"right":    n.Right.ToMap(),
	}
}

// UnaryNode is the logical-not operation ("!" operand).
type UnaryNode struct {
	Operator string
	Operand  Node
}

// ToMap serializes the Unary node and its operand to the shared-fixture AST
// shape {type, operator, operand}.
func (n *UnaryNode) ToMap() map[string]any {
	return map[string]any{
		"type":     "Unary",
		"operator": n.Operator,
		"operand":  n.Operand.ToMap(),
	}
}

// InNode is value in/not-in list. List holds Literal nodes.
type InNode struct {
	Negated bool
	Value   Node
	List    []Node
}

// ToMap serializes the In node and its list to the shared-fixture AST shape
// {type, negated, value, list}.
func (n *InNode) ToMap() map[string]any {
	list := make([]any, len(n.List))
	for i, item := range n.List {
		list[i] = item.ToMap()
	}
	return map[string]any{
		"type":    "In",
		"negated": n.Negated,
		"value":   n.Value.ToMap(),
		"list":    list,
	}
}

// PathSegment is one Path step. Type is "identifier", "wildcard", or "index".
type PathSegment struct {
	Type  string
	Value string // identifier name
	Index int    // index value
}

// ToMap serializes the segment to the JS segment shape (index value as number).
func (s PathSegment) ToMap() map[string]any {
	switch s.Type {
	case "identifier":
		return map[string]any{"type": "identifier", "value": s.Value}
	case "wildcard":
		return map[string]any{"type": "wildcard"}
	case "index":
		return map[string]any{"type": "index", "value": s.Index}
	default:
		return map[string]any{"type": s.Type}
	}
}

// PathNode is a field reference. Relative true => "."/".."; LevelsUp is parents
// to ascend (".x"=0, "..x"=1). Segments are the dotted path after the prefix.
type PathNode struct {
	Relative bool
	LevelsUp int
	Segments []PathSegment
}

// ToMap serializes the Path node and its segments to the shared-fixture AST
// shape {type, relative, levelsUp, segments}.
func (n *PathNode) ToMap() map[string]any {
	segs := make([]any, len(n.Segments))
	for i, s := range n.Segments {
		segs[i] = s.ToMap()
	}
	return map[string]any{
		"type":     "Path",
		"relative": n.Relative,
		"levelsUp": n.LevelsUp,
		"segments": segs,
	}
}

// LiteralNode is a constant. ValueType is "string"|"number"|"boolean"|"null".
type LiteralNode struct {
	ValueType string
	Value     any
}

// ToMap serializes the Literal node to the shared-fixture AST shape {type,
// valueType, value}.
func (n *LiteralNode) ToMap() map[string]any {
	return map[string]any{
		"type":      "Literal",
		"valueType": n.ValueType,
		"value":     n.Value,
	}
}

// GroupNode is a parenthesized expression.
type GroupNode struct {
	Expression Node
}

// ToMap serializes the Group node and its inner expression to the
// shared-fixture AST shape {type, expression}.
func (n *GroupNode) ToMap() map[string]any {
	return map[string]any{
		"type":       "Group",
		"expression": n.Expression.ToMap(),
	}
}

File: packages/validator-rust/src/expr/ast.rs
//! CRUDUI expression AST nodes (expressions.md §4, JS `ConditionParser.ts` parity).
//!
//! `to_value()` emits the canonical fixture shape, byte-matching the JS reference
//! field names: `Ternary{condition,trueValue,falseValue}`, `Binary{operator,left,
//! right}`, `Unary{operator,operand}`, `In{negated,value,list}`, `Path{relative,
//! levelsUp,segments}`, `Literal{valueType,value}`, `Group{expression}`. The
//! JS/Go `position` field is DELIBERATELY ABSENT — Rust holds no position, so the
//! 4-language AST contract excludes it; including it would break cross-language
//! fixture equality.

use serde_json::{json, Value};

/// A literal's value kind plus its concrete value (number kept as int or float so
/// the fixture's integer literals stay integers when serialized).
#[derive(Debug, Clone, PartialEq)]
pub enum LiteralValue {
    /// `valueType:"string"`.
    Str(String),
    /// `valueType:"number"`, integral.
    Int(i64),
    /// `valueType:"number"`, fractional.
    Float(f64),
    /// `valueType:"boolean"`.
    Bool(bool),
    /// `valueType:"null"`.
    Null,
}

impl LiteralValue {
    fn value_type(&self) -> &'static str {
        match self {
            LiteralValue::Str(_) => "string",
            LiteralValue::Int(_) | LiteralValue::Float(_) => "number",
            LiteralValue::Bool(_) => "boolean",
            LiteralValue::Null => "null",
        }
    }

    fn to_value(&self) -> Value {
        match self {
            LiteralValue::Str(s) => Value::String(s.clone()),
            LiteralValue::Int(n) => json!(n),
            LiteralValue::Float(f) => json!(f),
            LiteralValue::Bool(b) => Value::Bool(*b),
            LiteralValue::Null => Value::Null,
        }
    }
}

/// A single path segment: identifier (field name), wildcard (`*`), or index
/// (numeric array index). The index value serializes as a number to match the JS
/// reference segment shape (`{type:'index', value:number}`).
#[derive(Debug, Clone, PartialEq)]
pub enum PathSegment {
    /// A named-field segment.
    Identifier(String),
    /// A wildcard (`*`) segment matching any array index.
    Wildcard,
    /// A fixed numeric array-index segment.
    Index(i64),
}

impl PathSegment {
    fn to_value(&self) -> Value {
        match self {
            PathSegment::Identifier(v) => json!({ "type": "identifier", "value": v }),
            PathSegment::Wildcard => json!({ "type": "wildcard" }),
            PathSegment::Index(i) => json!({ "type": "index", "value": i }),
        }
    }
}

/// An expression AST node (expressions.md §4). CRUDUI-only; never reuse the legacy
/// `crate::condition_parser` model (R7 parallel run).
#[derive(Debug, Clone, PartialEq)]
pub enum Node {
    /// `condition ? trueValue : falseValue` (right-associative).
    Ternary {
        /// The boolean condition.
        condition: Box<Node>,
        /// Branch evaluated when the condition is truthy.
        true_value: Box<Node>,
        /// Branch evaluated when the condition is falsy.
        false_value: Box<Node>,
    },
    /// Logical (`&& ||`) or comparison (`== != > >= < <=`) operation.
    Binary {
        /// The operator spelling.
        operator: String,
        /// Left operand.
        left: Box<Node>,
        /// Right operand.
        right: Box<Node>,
    },
    /// `! operand`.
    Unary {
        /// The operator spelling (`!`).
        operator: String,
        /// The negated operand.
        operand: Box<Node>,
    },
    /// `value [not] in list` membership test.
    In {
        /// Whether the test is negated (`not in`).
        negated: bool,
        /// The probed value.
        value: Box<Node>,
        /// Candidate Literal nodes.
        list: Vec<Node>,
    },
    /// A path reference.
    Path {
        /// `false` → resolved from the data root; `true` → relative to the field.
        relative: bool,
        /// Additional parent ascents beyond the current field's parent.
        levels_up: i64,
        /// Ordered path segments.
        segments: Vec<PathSegment>,
    },
    /// A literal value.
    Literal(LiteralValue),
    /// A parenthesized group wrapping an inner expression.
    Group(Box<Node>),
}

impl Node {
    /// Serialize to the canonical fixture AST shape (no `position`).
    pub fn to_value(&self) -> Value {
        match self {
            Node::Ternary {
                condition,
                true_value,
                false_value,
            } => json!({
                "type": "Ternary",
                "condition": condition.to_value(),
                "trueValue": true_value.to_value(),
                "falseValue": false_value.to_value(),
            }),
            Node::Binary {
                operator,
                left,
                right,
            } => json!({
                "type": "Binary",
                "operator": operator,
                "left": left.to_value(),
                "right": right.to_value(),
            }),
            Node::Unary { operator, operand } => json!({
                "type": "Unary",
                "operator": operator,
                "operand": operand.to_value(),
            }),
            Node::In {
                negated,
                value,
                list,
            } => json!({
                "type": "In",
                "negated": negated,
                "value": value.to_value(),
                "list": list.iter().map(Node::to_value).collect::<Vec<_>>(),
            }),
            Node::Path {
                relative,
                levels_up,
                segments,
            } => json!({
                "type": "Path",
                "relative": relative,
                "levelsUp": levels_up,
                "segments": segments.iter().map(PathSegment::to_value).collect::<Vec<_>>(),
            }),
            Node::Literal(lit) => json!({
                "type": "Literal",
                "valueType": lit.value_type(),
                "value": lit.to_value(),
            }),
            Node::Group(expr) => json!({
                "type": "Group",
                "expression": expr.to_value(),
            }),
        }
    }
}
